src/controller/java/OTAProviderDelegateBridge.cpp

```diff
@@ -0,0 +1,409 @@
+        }
+
+        JniUtfString jniSoftwareVersionString(env, jSoftwareVersionString);
+
+        if (jniSoftwareVersionString.c_str() != nullptr)
+        {
+            response.softwareVersionString.SetValue(jniSoftwareVersionString.charSpan());
+        }
+
+        JniUtfString jniFilePath(env, jFilePath);
+
+        GenerateUpdateToken(mToken, kUpdateTokenLen);
+
+        response.updateToken.SetValue(chip::ByteSpan(mToken, kUpdateTokenLen));
+        response.userConsentNeeded.SetValue(0);
+
+        err = mBdxOTASender->PrepareForTransfer(fabricIndex, nodeId);
+        if (CHIP_NO_ERROR != err)
+        {
+            // Handle busy error separately as we have a query image response status that maps to busy
+            if (err == CHIP_ERROR_BUSY)
+            {
+                ChipLogError(Controller, "Responding with Busy due to being in the middle of handling another BDX transfer");
+                Commands::QueryImageResponse::Type errorResponse;
+                errorResponse.status = OTAQueryStatus::kBusy;
+                errorResponse.delayedActionTime.SetValue(kDelayedActionTimeSeconds);
+                commandObj->AddResponse(cachedCommandPath, errorResponse);
+                // We do not reset state when we get the busy error because that means we are locked in a BDX transfer
+                // session with another requestor when we get this query image request. We do not want to interrupt the
+                // ongoing transfer instead just respond to the second requestor with a busy status and a delayedActionTime
+                // in which the requestor can retry.
+                ExitNow(errorStatus = static_cast<uint8_t>(errorResponse.status));
+            }
+            LogErrorOnFailure(err);
+            commandObj->AddStatus(cachedCommandPath, StatusIB(err).mStatus);
+            // We need to reset state here to clean up any initialization we might have done including starting the BDX
+            // timeout timer while preparing for transfer if any failure occurs afterwards.
+            mBdxOTASender->ResetState();
+            ExitNow(errorStatus = static_cast<uint8_t>(OTAQueryStatus::kNotAvailable));
+        }
+        err = bdx::MakeURI(ourNodeId.GetNodeId(), jniFilePath.c_str() != nullptr ? jniFilePath.charSpan() : CharSpan(), uri);
+        if (CHIP_NO_ERROR != err)
+        {
+            LogErrorOnFailure(err);
+            commandObj->AddStatus(cachedCommandPath, StatusIB(err).mStatus);
+            mBdxOTASender->ResetState();
+            ExitNow(errorStatus = static_cast<uint8_t>(OTAQueryStatus::kNotAvailable));
+        }
+        response.imageURI.SetValue(uri);
+        commandObj->AddResponse(cachedCommandPath, response);
+    }
+    return;
+
+exit:
+    ChipLogError(Controller, "OTA Query Failure : %u, %" CHIP_ERROR_FORMAT, errorStatus, err.Format());
+    sendOTAQueryFailure(errorStatus);
```

Comment:
when error happens, should we commandObj->AddResponse with bad status?

Reply:
Understand. I modified.